SLHCUpgradeSimulations/Geometry/test/Phase2PixelNtuple.cc

```diff
@@ -0,0 +1,908 @@
+  pixeltreeOnTrack_->Branch("hrow",	&recHit_.hrow   	 ,"hrow/F");		   
+  pixeltreeOnTrack_->Branch("hcol",	&recHit_.hcol   	 ,"hcol/F");
+  pixeltreeOnTrack_->Branch("tx",	&recHit_.tx		 ,"tx/F");	   
+  pixeltreeOnTrack_->Branch("ty",	&recHit_.ty		 ,"ty/F");	   
+  pixeltreeOnTrack_->Branch("tz",	&recHit_.tz		 ,"tz/F");	   
+  pixeltreeOnTrack_->Branch("theta",	&recHit_.theta		 ,"theta/F" );	   
+  pixeltreeOnTrack_->Branch("phi",	&recHit_.phi		 ,"phi/F"    );	
+  pixeltreeOnTrack_->Branch("nRowsInDet", &recHit_.nRowsInDet	 ,"nRowsInDet/I");	
+  pixeltreeOnTrack_->Branch("nColsInDet", &recHit_.nColsInDet	 ,"nColsInDet/I");
+  pixeltreeOnTrack_->Branch("pitchx",     &recHit_.pitchx	 ,"pitchx/F");
+  pixeltreeOnTrack_->Branch("pitchy",     &recHit_.pitchy	 ,"pitchy/F");	  	  
+  pixeltreeOnTrack_->Branch("thickness",  &recHit_.thickness	 ,"thickness/F");	   
+  pixeltreeOnTrack_->Branch("cotAlphaFromDet",	&recHit_.cotAlphaFromDet  ,"cotAlphaFromDet/F");	   
+  pixeltreeOnTrack_->Branch("cotBetaFromDet",	&recHit_.cotBetaFromDet	  ,"cotBetaFraomDet/F");	   
+
+  pixeltreeOnTrack_->Branch("DgN",	&recHit_.fDgN		 ,"DgN/I"    );		   
+  pixeltreeOnTrack_->Branch("DgRow",	 recHit_.fDgRow		 ,"DgRow[DgN]/I"  );	   
+  pixeltreeOnTrack_->Branch("DgCol",	 recHit_.fDgCol		 ,"DgCol[DgN]/I"   ); 
+  pixeltreeOnTrack_->Branch("DgDetId",	 recHit_.fDgDetId	 ,"DgDetId[DgN]/I" );  
+  pixeltreeOnTrack_->Branch("DgAdc",	 recHit_.fDgAdc		 ,"DgAdc[DgN]/F"   ); 
+  pixeltreeOnTrack_->Branch("DgCharge",  recHit_.fDgCharge	 ,"DgCharge[DgN]/F");  
+}
+ 
+// Functions that gets called by framework every event
+void Phase2PixelNtuple::analyze(const edm::Event& e, const edm::EventSetup& es)
+{
+  //Retrieve tracker topology from geometry
+  edm::ESHandle<TrackerTopology> tTopoHandle;
+  es.get<TrackerTopologyRcd>().get(tTopoHandle);
+  const TrackerTopology* const tTopo = tTopoHandle.product();
+ 
+  // geometry setup
+  edm::ESHandle<TrackerGeometry>        geometry;
+ 
+  es.get<TrackerDigiGeometryRecord>().get(geometry);
+  const TrackerGeometry*  theGeometry = &(*geometry);
+  
+  std::vector<PSimHit> matched;
+  std::vector<PSimHit>::const_iterator closest_simhit;
+
+  edm::Handle<SiPixelRecHitCollection> recHitColl;
+  //e.getByLabel( src_, recHitColl);
+  e.getByToken(pixelRecHits_token, recHitColl);
+  // for finding matched simhit
+  TrackerHitAssociator associate( e, trackerHitAssociatorConfig_);
+ 
+  //  std::cout << " Step A: Standard RecHits found " << (recHitColl.product())->dataSize() << std::endl;
+  if((recHitColl.product())->dataSize() > 0) {
+    SiPixelRecHitCollection::const_iterator recHitIdIterator      = (recHitColl.product())->begin();
+    SiPixelRecHitCollection::const_iterator recHitIdIteratorEnd   = (recHitColl.product())->end();
+ 
+    std::string detname ;
+
+    evt_.init();
+    fillEvt(e);
+
+    // Loop over Detector IDs
+    for ( ; recHitIdIterator != recHitIdIteratorEnd; recHitIdIterator++) {
```

Comment:
use range-based loop here?

Reply:
ok for this one I gave up, it would require more changes in the code , some maybe I would leave it there as it is - All the other range-loop requests (and other comments) have been adressed now- Thanks

Reply:
@boudoul - perhaps, removing extra variables from a loop would help:
```
diff --git a/SLHCUpgradeSimulations/Geometry/test/Phase2PixelNtuple.cc b/SLHCUpgradeSimulations/Geometry/test/Phase2PixelNtuple.cc
index 2208c643bdd..89d3ac033d3 100644
--- a/SLHCUpgradeSimulations/Geometry/test/Phase2PixelNtuple.cc
+++ b/SLHCUpgradeSimulations/Geometry/test/Phase2PixelNtuple.cc
@@ -310,8 +310,6 @@ void Phase2PixelNtuple::analyze(const edm::Event& e, const edm::EventSetup& es)
   TrackerHitAssociator associate( e, trackerHitAssociatorConfig_);
  
   if((recHitColl.product())->dataSize() > 0) {
-    SiPixelRecHitCollection::const_iterator recHitIdIterator      = (recHitColl.product())->begin();
-    SiPixelRecHitCollection::const_iterator recHitIdIteratorEnd   = (recHitColl.product())->end();
  
     std::string detname ;
 
@@ -319,8 +317,8 @@ void Phase2PixelNtuple::analyze(const edm::Event& e, const edm::EventSetup& es)
     fillEvt(e);
 
     // Loop over Detector IDs
-    for ( ; recHitIdIterator != recHitIdIteratorEnd; recHitIdIterator++) {
-      SiPixelRecHitCollection::DetSet detset = *recHitIdIterator;
+    for ( auto recHitIdIterator : *(recHitColl.product())) {
+      SiPixelRecHitCollection::DetSet detset = recHitIdIterator;
  
       if( detset.empty() ) continue;
       DetId detId = DetId(detset.detId()); // Get the Detid object
@@ -328,17 +326,13 @@ void Phase2PixelNtuple::analyze(const edm::Event& e, const edm::EventSetup& es)
       const GeomDet* geomDet( theGeometry->idToDet(detId) );
        
       // Loop over rechits for this detid
-      SiPixelRecHitCollection::DetSet::const_iterator rechitRangeIteratorBegin = detset.begin();
-      SiPixelRecHitCollection::DetSet::const_iterator rechitRangeIteratorEnd   = detset.end();
-      SiPixelRecHitCollection::DetSet::const_iterator iterRecHit;
-      for ( iterRecHit = rechitRangeIteratorBegin; 
-            iterRecHit != rechitRangeIteratorEnd; ++iterRecHit) {
+      for ( auto iterRecHit : detset ) {
        // get matched simhit
        matched.clear();
-       matched = associate.associateHit(*iterRecHit);
+       matched = associate.associateHit(iterRecHit);
        if ( !matched.empty() ) {
          float closest = 9999.9;
-         LocalPoint lp = iterRecHit->localPosition();
+         LocalPoint lp = iterRecHit.localPosition();
          float rechit_x = lp.x();
          float rechit_y = lp.y();
          //loop over simhits and find closest
@@ -382,7 +376,7 @@ void Phase2PixelNtuple::analyze(const edm::Event& e, const edm::EventSetup& es)
          int num_simhit = matched.size();
          recHit_.init();
          fillPRecHit(  detid_db, subid,layer_num,ladder_num,module_num,disk_num,blade_num,panel_num,side_num,
-                       iterRecHit, num_simhit, closest_simhit, geomDet );
+                       &iterRecHit, num_simhit, closest_simhit, geomDet );
          pixeltree_->Fill();
        }
       } // end of rechit loop

```

Reply:
Thanks a lot @ianna !!! I applied your suggestion .